csrc/executor.cpp

```diff
@@ -1892,14 +1924,41 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
     inputBytesProcessed(args);
   }
 
-  if (!hasCompiledKernel()) {
+  if (isExpressionEvaluated(fusion())) {
     outputs = evaluateFusionOutputs(args, outputs, expr_eval);
     if (measure_kernel_time) {
       outputBytesProcessed(outputs);
     }
     return outputs;
   }
 
+  if (host_ir_container_ != nullptr) {
+    if (outputs.empty()) {
+      std::vector<GlobalBufferInfo> output_info = getOutputBufferInfo(
+          args, expr_eval, PrimDataType::Int, host_ir_container_.get());
+      outputs = allocateOutputs(
+          host_ir_container_.get(), output_info, options_.device, expr_eval);
+    }
+    for (Expr* e : host_ir_container_->topLevelExprs()) {
```

Comment:
I debated that myself too. Eventually, FusionExecutor won't be responsible to post communication or evaluate ATen. So I tried to avoid dependency on HostIrExecutor just for running this communication-only for-loop. This way, you can potentially iterate faster on HostIrExecutor.

Reply:
alright, sounds good, thanks